### ISSUE TITLE
Download binaries from server

### ISSUE THREAD

**Issue by ruslo** [ruslo/hunter#75]
It should be quite simple task after implementation of [cache](https://github.com/ruslo/hunter/issues/74). I.e. merge several `${HUNTER_ROOT}/_Base/Cache` directories from Jenkins server and make it public. `hunter_download` can use `file(DOWNLOAD` to check meta information and on cache hit download archive with dependencies instead of building them from sources.

**Comment by ruslo** [comment]
Update note: we can start with Travis/AppVeyor

**Comment by ruslo** [comment]
Downloading binaries from cache feature available since [v0.14.0](https://github.com/ruslo/hunter/releases/tag/v0.14.0) version, see [this overview](https://github.com/ruslo/hunter/wiki/Uploading-binaries-to-server).

New variables introduced:
* [HUNTER_DISABLE_BUILDS](https://github.com/ruslo/hunter/wiki/usr.variables#hunter_disable_builds)
* [HUNTER_CACHE_SERVERS](https://github.com/ruslo/hunter/wiki/usr.variables#hunter_cache_servers)
* [HUNTER_USE_CACHE_SERVERS](https://github.com/ruslo/hunter/wiki/usr.variables#hunter_use_cache_servers)

Default cache server:
* https://github.com/ingenue/hunter-cache
* https://github.com/ingenue/hunter-cache/releases/tag/cache

Build with GTest package downloaded from server:
* https://ci.appveyor.com/project/ruslo/hunter-simple/build/1.0.8
* https://travis-ci.org/forexample/hunter-simple/builds/116322978

Includes next toolchains (see [toolchain list available for Travis/AppVeyor](https://github.com/travis-ci-tester/toolchain-table)):

Linux/OSX:
* android-ndk-r10e-api-19-armeabi-v7a-neon
* clang-libstdcxx
* gcc
* libcxx
* analyze

OSX:
* ios-nocodesign
* osx-10-9
* xcode

Visual Studio:
* vs-9-2008
* vs-10-2010
* vs-11-2012
* vs-12-2013
* vs-12-2013-win64
* vs-12-2013-xp
* vs-14-2015

Windows:
* nmake-vs-12-2013
* nmake-vs-12-2013-win64
* ninja-vs-12-2013-win64
* mingw

Sanitizers
* sanitize-address
* sanitize-leak
* sanitize-thread

*note: I never tested sanitizers in fact, just checked that flags passed correctly :)*

To use cache the environment should fit environment from Travis/AppVeyor server. It will be easy to have same `config-id` and `hunter-id` but I can predict that the main issue will be `toolchain-id` since there are a lot of versions of different compilers. Though it's not a problem for the Travis/AppVeyor since environment is the same.

For now only GTest package added. Will add more step-by-step. Testing flow is next now:
```
`develop` (ruslo/hunter, unit testing)
  -> `pkg.template` (ingenue/hunter, dummy template)
    -> `pkg.foo` (ingenue/hunter, deep per package testing)
      -> `upload.foo` (ingenue/hunter, package upload)
```

See `upload.gtest` builds:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.255
* https://travis-ci.org/ingenue/hunter/builds/116212020

**Comment by ruslo** [comment]
> cache feature available since v0.14.0 version

Bug was found. In some cases dependency registration doesn't work. I've marked [v0.14.0](https://github.com/ruslo/hunter/releases/tag/v0.14.0) as a `buggy` version. Fixed in [v0.14.1](https://github.com/ruslo/hunter/releases/tag/v0.14.1). Since it may affect any package I have to update toolchain-id generation, effectively it means that all cache of old package will not be reused, sorry :(

Some cool facts I want to note:
* `toolchain-id` for Android (`Toolchain-ID: 8c15f91`) is the same for OSX and Linux hosts. It means they share same cache. It does work as expected, no surprises. Just want to note :)
 * https://travis-ci.org/ingenue/hunter/jobs/116842920
 * https://travis-ci.org/ingenue/hunter/jobs/116842926
* List of uploaded packages so far:
 * https://github.com/ingenue/hunter/branches/all?utf8=%E2%9C%93&query=upload
* Uploaded components and supported toolchains can be found in matrix, e.g. Boost:
 * https://travis-ci.org/ingenue/hunter/builds/116593022
* Upload from local environment for been reused in CI. Builds that terminated by timeout can be built locally (again, CI environment should match local environment precisely enough to produce same `toolchain-id`). As a PoC I've uploaded OpenCV binaries for `ios` and it works on Travis CI fine:
 * https://travis-ci.org/ingenue/hunter/jobs/116890222
* Build/upload on CI then reuse locally. Binaries uploaded to CI can be reused locally. I've run local build of GTest and binaries downloaded from server cache (environment note same as above).
* Performance results of building OpenCV example (dependencies `Eigen`, `PNG`, `TIFF`, `ZLIB` and obviously `OpenCV`):
 * Travis:
  * `1 min 42 sec` for `osx + libcxx`
  * `4 min 36 sec` for `osx + android` (it includes downloading/unpacking 3GB of Android NDK)
  * https://travis-ci.org/ingenue/hunter/builds/116890210
 * AppVeyor:
  * from `2 min 8 sec` to `3 min 13 sec`
  * https://ci.appveyor.com/project/ingenue/hunter/build/1.0.282
